### DISCUSSION TITLE
Post Processing With Multiple Profiles?

### DISCUSSION THREAD

**Discussion by youngpm** [onthegomap/planetiler#786] [Q&A]
Hello!  I was curious if it is possible to "post-process" multiple Profiles together before they are written to an MVT; in other words consider all the features in all layers that are going into the MVT.  For example, if I had "buildings" and "roads"  Profiles and wanted to tag a building with the closest road to it in the tile.  Is such a thing possible?  

I've seen the [post-processing](https://github.com/onthegomap/planetiler/blob/main/planetiler-core/src/main/java/com/onthegomap/planetiler/Profile.java#L107) in the Profile interface but it wasn't obvious to me how to do any cross-Profile interaction.  

Thanks!

**Comment by msbarry** [comment]
The API doesn't currently support this but it wouldn't be too hard to add.  What do you think the right API for this would look like? something like `(zoom, map<layer, list of features>) -> new map<layer, list of features>)` ? or `(zoom, list of features that include layer) -> new list of features that include layer` or something different?

**Reply by youngpm** [reply]
Thanks for the info!  

I'm not sure on the best interface (very new to planetiler!); in Tilezen's tilequeue/vector_datasource post processing, they pass a [Context](https://github.com/tilezen/tilequeue/blob/7dd1998e2c5c58182c9db49271b70cac4eab5c9d/tilequeue/process.py#L85-L122) that holds all the layers and their features plus a little geo info on the tile and update the result, so very similar to your first option.

**Reply by msbarry** [reply]
> new layers to be created from processing existing ones

I hadn't even considered that but it makes sense. So with tilezen api you provide functions that return a new layer or replacement for existing layer? or do they mutate the context? Passing in some more data on what tile you're processing makes sense too.

**Reply by youngpm** [reply]
Yup in tilezen, you configure the post-processing via this [queries.yaml](https://github.com/tilezen/vector-datasource/blob/master/queries.yaml#L353) file, a big list of python functions + parameters to apply to the output layers.  Each post-processing function modifies existing layers or can create new ones, and the result is passed on to the next post-processing function.  It is all mutable; if you change the features in a layer that'll be reflected in the output.  That part can be confusing though!

**Reply by msbarry** [reply]
Hmm, yeah it's a bit confusing to either mutate the inputs or return a new output... maybe it's just sufficient to have one post-process method that takes a similar context to what tilezen provides (I think `TileCoord` and `Map<layer, features>` would do it since `TileCoord` has zoom and some helper methods to get bounds/center)? Then the profile could mutate those, and probably have a fallback that calls the old `postProcess(layer)` method for each layer in the context as well for compatibility.

**Reply by bdon** [reply]
The one class to one MVT layer output is a nice separation of logic-as is; I have a branch that does Tilezen-style intercutting with this separation by including Roads in the building layer, performing geometry operations in postProcess, and then throwing away the roads at the end of postProcess so the never make it into the tiles. This could be expensive in time + temporary disk space though.

**Reply by msbarry** [reply]
The class per layer separation is from [ForwardingProfile](https://github.com/onthegomap/planetiler/blob/14b217d6f6bb1ac205b7905380fd6a7d630c02c1/planetiler-core/src/main/java/com/onthegomap/planetiler/ForwardingProfile.java#L129), we could keep that delegation behavior but have the Profile method that it implements be more flexible and process all the layers at once?

**Reply by msbarry** [reply]
I also wonder if there's a way to model dependencies between layers that are merged/created/dropped from other layers for classes in forwarding profile? Right now when you run with `--only-layers=water,building` it only loads the class for water and building but if they depended on transportation for post-processing then ideally we'd want it to automatically load transportation class too but omit the layer from the output.

**Reply by bdon** [reply]
Most of the Tilezen operations should work on the "visual" end-state output of the layer, after all merging/simplification, so expressing a dependency between layers would be better than doing redundant work by loading the features twice like I did above.

**Reply by msbarry** [reply]
I put out a prototype for this in #802. It adds a new method to `Profile` you can override to post-process features across layers.

```java
  default Map<String, List<VectorTile.Feature>> postProcessTileFeatures(TileCoord tileCoord,
    Map<String, List<VectorTile.Feature>> layers) throws GeometryException {
    return layers; // default: do nothing
  }
```

Thoughts? If this looks good, we can hook it into `ForwardingProfile` to let complex profiles stack the transformation functions.

**Reply by youngpm** [reply]
Awesome, looks good to me!

**Reply by msbarry** [reply]
👍 just merged #802 so it should be possible now, will think some more about how to best handle splitting logic between classes in ForwardingProfile...
